src/gamestate/commands.cpp

```diff
@@ -4479,352 +4481,384 @@ void execute_notify_player_picks_nation(sys::state& state, dcon::nation_id sourc
-			break;
-		case command_type::delete_army:
-			execute_delete_army(state, c.source, c.data.army_movement.a);
-			break;
-		case command_type::delete_navy:
-			execute_delete_navy(state, c.source, c.data.navy_movement.n);
-			break;
-		case command_type::designate_split_regiments:
-			execute_mark_regiments_to_split(state, c.source, c.data.split_regiments.regs);
-			break;
-		case command_type::designate_split_ships:
-			execute_mark_ships_to_split(state, c.source, c.data.split_ships.ships);
-			break;
-		case command_type::naval_retreat:
-			execute_retreat_from_naval_battle(state, c.source, c.data.naval_battle.b);
-			break;
-		case command_type::land_retreat:
-			execute_retreat_from_land_battle(state, c.source, c.data.land_battle.b);
-			break;
-		case command_type::start_crisis_peace_offer:
-			execute_start_crisis_peace_offer(state, c.source, c.data.new_offer.is_concession);
-			break;
-		case command_type::invite_to_crisis:
-			execute_invite_to_crisis(state, c.source, c.data.crisis_invitation);
-			break;
-		case command_type::add_wargoal_to_crisis_offer:
-			execute_add_to_crisis_peace_offer(state, c.source, c.data.crisis_invitation);
-			break;
-		case command_type::send_crisis_peace_offer:
-			execute_send_crisis_peace_offer(state, c.source);
-			break;
-		case command_type::change_admiral:
-			execute_change_admiral(state, c.source, c.data.new_admiral.a, c.data.new_admiral.l);
-			break;
-		case command_type::change_general:
-			execute_change_general(state, c.source, c.data.new_general.a, c.data.new_general.l);
-			break;
-		case command_type::toggle_mobilization:
-			execute_toggle_mobilization(state, c.source);
-			break;
-		case command_type::give_military_access:
-			execute_give_military_access(state, c.source, c.data.diplo_action.target);
-			break;
-		case command_type::set_rally_point:
-			execute_set_rally_point(state, c.source, c.data.rally_point.location, c.data.rally_point.naval, c.data.rally_point.enable);
-			break;
-		case command_type::save_game:
-			execute_save_game(state, c.source, c.data.save_game.and_quit);
-			break;
-		case command_type::cancel_factory_building_construction:
-			execute_cancel_factory_building_construction(state, c.source, c.data.start_factory_building.location, c.data.start_factory_building.type);
-			break;
-		case command_type::disband_undermanned:
-			execute_disband_undermanned_regiments(state, c.source, c.data.army_movement.a);
-			break;
-		case command_type::even_split_army:
-			execute_evenly_split_army(state, c.source, c.data.army_movement.a);
-			break;
-		case command_type::even_split_navy:
-			execute_evenly_split_navy(state, c.source, c.data.navy_movement.n);
-			break;
+void execute_advance_tick(sys::state& state, dcon::nation_id source, sys::checksum_key& checksum) {
+#ifndef NDEBUG
+	sys::checksum_key current = state.get_network_checksum();
+	if(!current.is_equal(checksum)) {
+#ifdef _WIN64
+		std::string msg = "Network has gotten out of sync: ";
```

Comment:
We can't be serializing the game state out every tick to generate a checksum. Maybe in debug, but we can't be sending it out every day in release

Reply:
My bad, it was meant to be debug only

Reply:
Wait a minute, there is literally a #ifndef NDEBUG above it

Reply:
Where is the code that *generates* the value, however?

Reply:
system_state.cpp:3387
```c++
sys::checksum_key state::get_network_checksum() {
	auto buffer = std::unique_ptr<uint8_t[]>(new uint8_t[sizeof_save_section(*this)]);

	dcon::load_record loaded = world.make_serialize_record_store_save();
	std::byte* start = reinterpret_cast<std::byte*>(buffer.get());
	world.serialize(start, loaded);

	auto buffer_position = reinterpret_cast<uint8_t*>(start);
	int32_t total_size_used = static_cast<int32_t>(buffer_position - buffer.get());

	checksum_key key;
	blake2b(&key, sizeof(key), buffer.get(), total_size_used, nullptr, 0);
	return key;
}
```

Reply:
wrap that n #ifndef NDEBUG with a case to return a dummy key in release -- this prevents any mistakes from happening there. Oh, and put the ifndef around the body of the function that puts the message into the queue too. Finally, make sure that it isn't going to be running in single player -- making single player performance that much worse in debug will kill me since I need it to find and fix problems.

Reply:
I ensured it is only called in multiplayer contexts

Reply:
or when loading/saving

Reply:
Just idiot proof things as much as possible, because calling that function by mistake anywhere it doesn't need to be called would be a huge performance drain -- that's 100s of MB of data that is being written out every time you call it.

Reply:
that would break savefiles and scenarios, if we return dummy keys that is